core/src/main/scala/cats/data/Catenable.scala

```diff
@@ -0,0 +1,263 @@
+    while (result eq null) {
+      c match {
+        case Empty =>
+          if (rights.isEmpty) {
+            result = None
+          } else {
+            c = rights.last
+            rights.trimEnd(1)
+          }
+        case Singleton(a) =>
+          val next =
+            if (rights.isEmpty) empty
+            else rights.reduceLeft((x, y) => Append(y, x))
+          result = Some(a -> next)
+        case Append(l, r) => c = l; rights += r
+      }
+    }
+    result
+  }
+
+  /** Returns true if there are no elements in this collection. */
+  def isEmpty: Boolean
+
+  /** Returns false if there are no elements in this collection. */
+  def nonEmpty: Boolean = !isEmpty
+
+  /** Concatenates this with `c` in O(1) runtime. */
+  final def ++[A2 >: A](c: Catenable[A2]): Catenable[A2] =
+    append(this, c)
+
+  /** Returns a new catenable consisting of `a` followed by this. O(1) runtime. */
+  final def cons[A2 >: A](a: A2): Catenable[A2] =
+    append(singleton(a), this)
+
+  /** Alias for [[cons]]. */
+  final def +:[A2 >: A](a: A2): Catenable[A2] =
+    cons(a)
+
+  /** Returns a new catenable consisting of this followed by `a`. O(1) runtime. */
+  final def snoc[A2 >: A](a: A2): Catenable[A2] =
+    append(this, singleton(a))
+
+  /** Alias for [[snoc]]. */
+  final def :+[A2 >: A](a: A2): Catenable[A2] =
+    snoc(a)
+
+  /** Applies the supplied function to each element and returns a new catenable. */
+  final def map[B](f: A => B): Catenable[B] =
+    foldLeft(empty: Catenable[B])((acc, a) => acc :+ f(a))
+
+  /** Applies the supplied function to each element and returns a new catenable from the concatenated results */
+  final def flatMap[B](f: A => Catenable[B]): Catenable[B] =
+    foldLeft(empty: Catenable[B])((acc, a) => acc ++ f(a))
+
+  /** Folds over the elements from left to right using the supplied initial value and function. */
+  final def foldLeft[B](z: B)(f: (B, A) => B): B = {
+    var result = z
+    foreach(a => result = f(result, a))
+    result
+  }
+
+  /** collect `B` from this for which `f` is defined **/
+  final def collect[B](f: PartialFunction[A, B]): Catenable[B] = {
+    val predicate = f.lift
+    foldLeft(Catenable.empty: Catenable[B]) { (acc, a) =>
+      predicate(a).fold(acc)(b => acc :+ b)
+    }
+  }
+
+  /**
+   * Yields to Some(a, Catenable[A]) with `a` removed where `f` holds for the first time,
+   * otherwise yields None, if `a` was not found
+   * Traverses only until `a` is found.
+   */
+  final def deleteFirst(f: A => Boolean): Option[(A, Catenable[A])] = {
+    @tailrec
+    def go(rem: Catenable[A], acc: Catenable[A]): Option[(A, Catenable[A])] =
+      rem.uncons match {
+        case Some((a, tail)) =>
+          if (!f(a)) go(tail, acc :+ a)
+          else Some((a, acc ++ tail))
+
+        case None => None
+      }
+    go(this, Catenable.empty)
+  }
+
+  /** Applies the supplied function to each element, left to right. */
+  private final def foreach(f: A => Unit): Unit = {
```

Comment:
I made this method private, though I think we could potentially just rename it to `unsafeForeach` instead? WDYT?